openvdb/points/AttributeSet.h

```diff
@@ -540,6 +558,82 @@ class OPENVDB_API AttributeSet::Descriptor
+
+/// @brief A convenience class that collates the Descriptor with metadata that
+/// is only available on each AttributeArray, such as whether the array is
+/// uniform or not. This is useful for querying and during construction of new
+/// AttributeSets.
+/// @note For performance reasons, this class uses copy-on-write for storing
+/// the Descriptor so any modification of the source Descriptor also requires
+/// a deep-copy for the data in this class to remain valid.
+class OPENVDB_API AttributeSet::Info {
+public:
+    struct OPENVDB_API Array {
+        Index stride = 1;
+        bool constantStride = true;
+        bool hidden = false;
+        bool transient = false;
+        bool group = false;
+        bool string = false;
+    }; // struct Array
+
+    /// Default constructor
+    Info() = default;
+
+    /// Construct a new Info class from the given Descriptor.
+    /// @param descriptorPtr the descriptor
+    /// @note Array metadata is default-initialized when using this constructor
+    explicit Info(const AttributeSet::Descriptor::Ptr& descriptorPtr);
+
+    /// Construct a new Info class from the given AttributeSet.
+    /// @param attributeSet the attribute set
+    /// @details This method extracts and stores Array metadata from the
+    /// AttributeArray objects stored within it
+    explicit Info(const AttributeSet& attributeSet);
+
+    /// @brief Return the descriptor
+    Descriptor& descriptor() { return *mDescriptor; }
+    /// @brief Return the descriptor
+    const Descriptor& descriptor() const { return *mDescriptor; }
+    /// @brief Return a shared pointer to the descriptor
+    Descriptor::Ptr descriptorPtr() const { return mDescriptor; }
+    /// @brief Return the number of array attributes
+    size_t size() const { return mDescriptor->size(); }
+
+    /// @brief Return an Array metadata object by name
+    /// @param name The name of the array
+    /// @throws LookupError if name does not exist
+    Array& arrayInfo(const Name& name);
+    /// @brief Return an Array metadata object by name
+    /// @param name The name of the array
+    /// @throws LookupError if name does not exist
+    const Array& arrayInfo(const Name& name) const;
+
+    /// @brief Return an Array metadata object by index
+    /// @param idx The index of the array
+    /// @throws LookupError if idx out of range
+    Array& arrayInfo(size_t idx);
+    /// @brief Return an Array metadata object by index
+    /// @param idx The index of the array
+    /// @throws LookupError if idx out of range
+    const Array& arrayInfo(size_t idx) const;
+
+    /// Placeholder method for merging another Info object into this one
+    /// @throws NotImplementedError
+    void merge(const AttributeSet::Info&);
+
+    /// @brief Print information for debugging
+    /// @param os the output stream to write to
+    void print(std::ostream& os = std::cout) const;
+
+private:
+    Descriptor::Ptr mDescriptor = std::make_shared<Descriptor>();
```

Comment:
At first glance, this class seems to me to be a "viewer" class which helps consolidate data from both the attribute set and descriptor. I question why the descriptor that's stored here is modifiable through the member methods. This introduces another layer of complexity. It seems that methods like `merge` would do better to return new instances of the descriptor.

Reply:
It does increase the complexity, but it's primarily done like this for performance reasons. Doing what you're suggesting doesn't scale to wanting to merge hundreds or even thousands of Info classes or Descriptors.

This class is serving two different use cases - however I consider the "viewer" use case to be mainly a side benefit, the real purpose is a composable description of what you want the AttributeSet to contain. (I did consider two different classes, one for each use case, but I preferred the simplicity of putting everything in one class). 

I would be open to changing the name if you think we can make this primary use-case clearer, `Info` is a pretty generic name.